syntax = "proto3";

option optimize_for = SPEED;

package todo;

service TodoService {
  rpc GetUsers (UsersRequest) returns (UsersList) {}
  rpc GetUser (User) returns (User) {}
  rpc GetUserTodos(User) returns (TodosList) {}
  rpc GetTodo(Todo) returns (Todo) {}
  rpc CreateTodo(Todo) returns (Todo) {}
  rpc CreateUser(User) returns (User) {}
}

// request messages
message UsersRequest {}

// messages
message User {
  int64 id = 1;
  string name = 2;
  string email = 3;
}

message UsersList {
  repeated User users = 1;
}

message Todo {
  int64 id = 1;
  int64 user_id = 2;
  string text = 3;
  Status status = 4;

  enum Status {
    UNKNOWN_STATUS = 0;
    NEW = 1;
    INPROGRESS = 2;
    DONE = 3;
  }
}

message TodosList {
  repeated Todo todos = 1;
}
